amt.go

```diff
@@ -16,56 +17,90 @@ import (
 
 var log = logging.Logger("amt")
 
-// MaxIndex is the maximum index for elements in the AMT. This is currently 1^63
-// (max int) because the width is 8. That means every "level" consumes 3 bits
-// from the index, and 63/3 is a nice even 21
-const MaxIndex = uint64(1<<internal.MaxIndexBits) - 1
+// MaxIndex is the maximum index for elements in the AMT. This MaxUint64-1 so we
+// don't overflow MaxUint64 when computing the length.
+const MaxIndex = math.MaxUint64 - 1
 
 type Root struct {
-	height int
-	count  uint64
+	bitWidth uint
+	height   int
+	count    uint64
 
 	node *node
 
 	store cbor.IpldStore
 }
 
-func NewAMT(bs cbor.IpldStore) *Root {
-	return &Root{
-		store: bs,
-		node:  new(node),
+func NewAMT(bs cbor.IpldStore, opts ...Option) (*Root, error) {
+	cfg := defaultConfig()
+	for _, opt := range opts {
+		if err := opt(cfg); err != nil {
+			return nil, err
+		}
 	}
+
+	return &Root{
+		bitWidth: cfg.bitWidth,
```

Comment:
Can we store the bitWidth in the top node of the AMT? If that requires a new struct to do so, I'd say it's worth it. We should strive for as much self-description as we can in these structures.

Reply:
Fair enough.

Reply:
@whyrusleeping as the author of this library, any thoughts here?

Reply:
💯 please do this

Reply:
@Stebalien are you blocking on this? If so is there a way I can help unblock?

Reply:
I'll ping why.

Reply:
eh, okay. we can do that now

Reply:
Ok, I've added it. However, I want to note that it does have limited usefulness:

* We can't rely on it in filecoin because we want to require a specific bitwidth. At best, we can use it in cross-version analysis tools.
* It won't help version across other changes like, e.g., fixing the AMT depth problem. That is, it'll make it easier for us to write tools that work across upgrades where we change the bitwidth, but not across upgrades where we change anything else.

Reply:
Thank you! I agree the benefit is greater for consumers outside the code that is actually writing the AMT, but I think that adds up to a lot over time (even though we won't directly learn of much of it).